docs: create document for diskpool io failure handling (#581)
openebs/website: docs/main/user-guides/replicated-storage-user-guide/replicated-pv-mayastor/advanced-operations/diskpool-io-failure-handling.md

## Configure Disk I/O Alert Thresholds

Replicated PV Mayastor allows you to configure DiskPool I/O monitoring thresholds using Helm values.

The following parameters control disk error and stall detection behavior:

| Parameter | Default Value | Description | Helm Path |
| :--- | :--- | :--- | :--- |
| `ioErrorThreshold` | `64` | Number of disk I/O errors allowed before the DiskPool is marked as `Suspected` | `mayastor.io_engine.pool.ioAlerts.errorThreshold` |
| `stallDeadline` | `io_engine.nvme.ioTimeout * 2` | Time interval after which incomplete I/O operations are treated as stalled | `mayastor.io_engine.pool.ioAlerts.stallDeadline` |
| `stallTransitionThreshold` | 3 | Number of intermittent stall transitions allowed before warning alerts are raised | `mayastor.io_engine.pool.ioAlerts.stallTransitionThreshold` |
| `stallTransitionWindow` | 3h | Time window used to track intermittent stall activity | `mayastor.io_engine.pool.ioAlerts.stallTransitionWindow` |

## View DiskPool Status

Use the following command to view DiskPool state, alerts, and conditions.

**Command**

```

> kubectl get diskpools -A

Reviewer: Should we not show examples here?

Reviewer: Also show using the plugin as well

Reviewer 2: Sure. @akashbc1989 can you please share a dsp yaml which has io_stall true and non zero error and stall counts.

Reviewer 3: @balaharish7  @abhilashshetty04  Here is the dsp cr yaml which has ioStalled true and non zero error and stall counts.

`k get dsp pool-2-on-node-0-474026 -n mayastor -o yaml
apiVersion: openebs.io/v1beta3
kind: DiskPool
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"openebs.io/v1beta3","kind":"DiskPool","metadata":{"annotations":{},"name":"pool-2-on-node-0-474026","namespace":"mayastor"},"spec":{"disks":["aio:///dev/mapper/sdb-timeout"],"node":"node-0-474026"}}
  creationTimestamp: "2026-06-02T06:18:12Z"
  finalizers:
  - openebs.io/diskpool-protection
  generation: 1
  name: pool-2-on-node-0-474026
  namespace: mayastor
  resourceVersion: "323743"
  uid: 7d433ae9-467e-408e-895e-b15571733fea
spec:
  disks:
  - aio:///dev/mapper/sdb-timeout
  node: node-0-474026
status:
  alertError: Critical (IoErrorExc,IoStalled)
  available: 8480882688
  available_q: 7.9 GiB
  capacity: 10724835328
  capacity_q: 10 GiB
  clusterSize: 4 MiB
  conditions:
  - lastTransitionTime: "2026-06-02T06:18:12Z"
    message: ""
    observedGeneration: 1
    reason: ""
    status: "True"
    type: PoolReady
  cr_state: Created
  diskCapacity: 10 GiB
  encrypted: false
  errorInfo:
    alerts:
      attention: []
      critical:
      - IoStalled
      notice: []
      status: Critical
      warning:
      - IoErrorExc
    ioErrorCount: 50399
    ioStallTransitionCount: 1
    ioStalled: true
  maxExpandableSize: 127.8 GiB
  pool_status: Suspected
  status: Suspected
  used: 2243952640
  used_q: 2.1 GiB`

Reviewer 2: @akashbc1989 , Can you please share screenshot instead?

:::

## DiskPool States and Alerts

Replicated PV Mayastor reports DiskPool health using pool states, alerts, and diagnostic information.

### Pool States

| State | Description |
| :--- | :--- |
| Online | Pool is operating normally |
| Suspected | Pool has warning-level disk or I/O issues |
| Faulted | Pool is inaccessible due to disk or I/O failures |
| Offline | Pool is unavailable because the node is offline or the disk is detached |
| Unknown | Pool state cannot be determined |

### Alert Status

| Alert Status | Description |
| :--- | :--- |
| Healthy | No disk I/O issues detected |
| Attention | Intermittent or low-frequency I/O stalls detected |
| Warning | Disk I/O error threshold exceeded or excessive intermittent stall transitions detected |
| Critical | Active stalled I/O operations detected |

## Hot-Removal Behavior

When a backing disk is detached becomes unavailable, Replicated PV Mayastor unloads the affected pool.

As part of this process:

- The DiskPool transitions to an Offline state
- The PoolReady condition becomes False
- Diagnostic information reports the detected failure reason

> The control plane probes the io-engine to detect whether the backing device has been reattached to the node. If the device becomes available again, Replicated PV Mayastor import the pool and restore normal operations.

Reviewer 2: ```suggestion
The control plane probes the io-engine to detect whether the backing device has been re-attached to the node. If the device becomes available again, Replicated PV Mayastor import the pool and restore normal operations.
```

Author: "re-attached" is accepted, but is increasingly considered outdated whereas "reattached" is Standard in most style guides.

## Disk I/O Error Handling

Replicated PV Mayastor tracks runtime disk I/O errors and exposes alert information for affected DiskPools.

If the number of I/O errors reported by a DiskPool exceeds the configured threshold, the DiskPool is marked as `Suspected`. Pools in the `Suspected` state are less likely to be selected for future volume placement operations.

The `ioErrorThreshold` parameter defines the number of allowed I/O errors before a DiskPool transitions to the `Suspected` state. By default, `ioErrorThreshold` is set to `64`. This parameter can be configured using Helm and applies globally to all DiskPools created in the cluster.

> 

Reviewer 2: Lets add:

DiskPool CR `status.error_info.io_error_count` exposes the current IO error count of the DiskPool.

## I/O Stall Detection and Intermittent Backend Failure Handling

Replicated PV Mayastor monitors DiskPools for stalled I/O operations.

When the I/O request is submitted to the pool backend and remains incomplete beyond the user-configured timeout interval (`stallDeadline`):

- The DiskPool is marked as stalled
- The DiskPool state transitions to Suspected
- The DiskPool alert status transitions to Critical
- The pool is avoided for future replica placement operations

I/O stalls can occur because of Storage device failures or unresponsive disks, Backend storage path instability, and Network disruptions affecting storage access.

> 

Reviewer 2: Lets add:

DiskPool `status.error_info.io_stalled` tells if the Pool is currently experiencing IO stalls and `status.error_info.io_stall_transition_count` shows the number of stall<->resume flips within the `stallTransitionWindow`

Replicated PV Mayastor allows you to configure DiskPool I/O monitoring thresholds using Helm values.

The following parameters control disk error and stall detection behavior:

| Parameter | Default Value | Description | Helm Path |
| :--- | :--- | :--- | :--- |
| `ioErrorThreshold` | `64` | Number of disk I/O errors allowed before the DiskPool is marked as `Suspected` | `mayastor.io_engine.pool.ioAlerts.errorThreshold` |
| `stallDeadline` | `io_engine.nvme.ioTimeout * 2` | Time interval after which incomplete I/O operations are treated as stalled | `mayastor.io_engine.pool.ioAlerts.stallDeadline` |
| `stallTransitionThreshold` | 3 | Number of intermittent stall transitions allowed before warning alerts are raised | `mayastor.io_engine.pool.ioAlerts.stallTransitionThreshold` |
| `stallTransitionWindow` | 3h | Time window used to track intermittent stall activity | `mayastor.io_engine.pool.ioAlerts.stallTransitionWindow` |

## View DiskPool Status

Use one of the following commands to view DiskPool state, alerts, and conditions.

**Command - View DiskPools Using Kubernetes Resources**

```
kubectl get diskpools -A
```

**Command - View Pools Using the OpenEBS Plugin**

```

> kubectl openebs mayastor get pools

Reviewer 2: ```suggestion
kubectl openebs mayastor get pool <pool-name> -n <namespace> -oyaml
```

Reviewer: Example output:
```
NAME               NODE          STATE     STATUS   ERROR   ALERTS                ENCRYPTED   CAPACITY   USED   AVAILABLE   DISK-CAPACITY   MAX-EXPANDABLE-SIZE
pool-kind-worker   kind-worker   Created   Online           Attention (IoError)   false       508 MiB    0 B    508 MiB     512 MiB         127.8 GiB
```

Author: Thanks @tiagolobocastro

```

## View DiskPool Details

Use the following command to inspect DiskPool conditions, alerts, and diagnostic information.

**Command**

```

> kubectl get dsp pool-2-on-node-0-474026 -n mayastor -o yaml

Reviewer 2: ```suggestion
kubectl get dsp <pool-name> -n <namespace> -o yaml
```

```

**Sample Output**

```
status:
  alertError: Critical (IoErrorExc,IoStalled)

  conditions:
  - status: "True"
    type: PoolReady

> 

Reviewer 2: Is blank line needed here?

Replicated PV Mayastor allows you to clear recorded disk I/O errors for a DiskPool after the underlying issue has been investigated and resolved. Clearing errors resets the `ioErrorCount` for the pool and removes any associated pool alerts that were generated as a result of the recorded I/O errors.

**Command**

```
kubectl openebs mayastor clear-errors pool <pool-id> -n <namespace>
```

:::important
Clearing errors resets the recorded I/O error count and associated alerts for the specified pool. This operation does not resolve underlying disk, device, or infrastructure issues. Ensure that the root cause has been addressed before clearing errors.
:::

## View DiskPool Status

Use one of the following commands to view DiskPool state, alerts, and conditions.

**Command - View DiskPools Using Kubernetes Resources**

```
kubectl get diskpools -A
```

**Command - View Pools Using the OpenEBS Plugin**

```
kubectl openebs mayastor get pools
```

## View DiskPool Details

Use the following command to inspect DiskPool conditions, alerts, and diagnostic information.

**Command**

```
kubectl get dsp pool-2-on-node-0-474026 -n mayastor -o yaml
```

**Sample Output**

```

> status:

Reviewer: tabled output:

```
 ID                DISKS                         MANAGED  NODE         STATUS  ALERTS               CAPACITY  ALLOCATED  AVAILABLE  COMMITTED  ENCRYPTED  DISK-CAPACITY  MAX-EXPANDABLE-SIZE  VOLUMES  SNAPSHOTS 
 pool-kind-worker  aio:///dev/kind-worker/lvol0  true     kind-worker  Online  Attention (IoError)  508 MiB   0 B        508 MiB    0 B        false      512 MiB        127.8 GiB            0        0 
```
